### PR TITLE
fix: add --version/--timeout flags to install and prevent kubectl os.Exit

### LINKED ISSUE #3631: CI Doctor - CI Failure - Run #9060 - Two New E2E Tests Failing (workload install flags + restore empty YAML)
## Summary

Two new E2E tests added in PR #3611 are failing in the merge queue due to code bugs: (1) `ksail workload install` missing `--version` and `--timeout` flags, and (2) `ksail cluster restore` producing "no objects passed to apply" due to an empty YAML after label injection.

## Failure Details

- **Run**: [23982636199](https://github.com/devantler-tech/ksail/actions/runs/23982636199)
- **Commit**: `c9ba53fa6311672f8eccdfd77093faca2276062b`
- **Trigger**: `merge_group` (PR #3611 — `feat: improve E2E test coverage across CLI commands`)

### 1. 🧪 System Test (Vanilla, Docker, true) — Job [69949409392](https://github.com/devantler-tech/ksail/actions/runs/23982636199/job/69949409392)

**Error:**
````
✗ unknown flag: --version
⚠️ workload install attempt 1 failed, retrying in 10s…
✗ unknown flag: --version
⚠️ workload install attempt 2 failed, retrying in 10s…
✗ unknown flag: --version
⚠️ workload install attempt 3 failed, retrying in 10s…
❌ workload install failed after 3 attempts
Process completed with exit code 1.
```

**Step:** `ksail workload install ksail-install-test oci://ghcr.io/stefanprodan/charts/podinfo --version=6.7.1 --wait --timeout=120s`

**Root Cause:** `pkg/cli/cmd/workload/install.go` only defines `--namespace`, `--create-namespace`, `--wait`, and `--atomic` flags. The `--version` and `--timeout` flags are missing, even though the underlying `helm.ChartSpec` struct supports them (`Version` and `Timeout` fields).

### 2. 🧪 Backup & Restore Test — Job [69949409299](https://github.com/devantler-tech/ksail/actions/runs/23982636199/job/69949409299)

**Error:**
```
Starting cluster restore...
   Input: /tmp/cluster-backup.tar.gz
   Policy: update
Extracting backup archive...
Backup metadata:
   Version: v1
   Resources: 233
Restoring cluster resources...
error: no objects passed to apply
Process completed with exit code 1.
````

**Root Cause:** The `ksail cluster restore` command's `injectRestoreLabels` function processes YAML files from the backup (233 resources), but when `sigsyaml.Unmarshal` + `sigsyaml.Marshal` transforms the YAML and writes to a temp file, the result seen by `kubectl apply -f` is empty or has no recognizable Kubernetes objects. The error "no objects passed to apply" is kubectl's response to an empty or non-K8s YAML. This triggers immediately on the first resource type (`customresourcedefinitions`), before any per-file warnings can be printed. The bug may be in the label injection round-trip losing the document content.

## Root Cause Analysis

Both failures are caused by new E2E tests in PR #3611 exposing pre-existing code deficiencies:

| Issue | Code Location | Missing / Bug |
|---|---|---|
| `--version` flag | `pkg/cli/cmd/workload/install.go:77-82` | `helm.ChartSpec.Version` field not wired to a CLI flag |
| `--timeout` flag | `pkg/cli/cmd/workload/install.go:47` | `helm.DefaultTimeout` hardcoded, no flag override |
| Restore empty YAML | `pkg/cli/cmd/cluster/restore.go:550-601` | `injectRestoreLabels` produces empty temp file for some resource types |

## Recommended Actions

- [ ] **Fix `install.go`**: Add `--version string` flag and wire it to `helm.ChartSpec.Version`. Add `--timeout duration` flag (defaulting to `helm.DefaultTimeout`) and wire it to `helm.ChartSpec.Timeout`.
  ```go
  flags.String("version", "", "specify a version constraint for the chart version to use")
  flags.Duration("timeout", helm.DefaultTimeout, "time to wait for any individual Kubernetes operation")
  ```
  Then read and apply in the RunE handler:
  ```go
  if version, _ := cmd.Flags().GetString("version"); version != "" {
      spec.Version = version
  }
  if timeout, _ := cmd.Flags().GetDuration("timeout"); timeout > 0 {
      spec.Timeout = timeout
  }
  ```

- [ ] **Debug `restore.go`**: Add debug logging to `injectRestoreLabels` to identify which resource type/file produces an empty result. Verify that `sigsyaml.Marshal(obj.Object)` after round-trip through `sigsyaml.Unmarshal` preserves `apiVersion` and `kind`. Consider using a simpler label injection that doesn't require full YAML round-trip (e.g., prepend label directly to YAML string via regex or header injection).

- [ ] **Add unit test for `injectRestoreLabels`** with a real CRD YAML to catch the empty-result regression.

## Reproduction Steps

```bash
# Reproduce the workload install issue:
cd /tmp && mkdir test-ksail && cd test-ksail
ksail cluster init
ksail cluster create
ksail workload install test-release oci://ghcr.io/stefanprodan/charts/podinfo --version=6.7.1
# Should fail with: unknown flag: --version

# Reproduce the restore issue (after cluster is running):
ksail cluster backup --output /tmp/test-backup.tar.gz
ksail cluster restore --input /tmp/test-backup.tar.gz --existing-resource-policy=update
# Should fail with: error: no objects passed to apply
```

## Prevention Strategies

- When adding new CI test steps that invoke `ksail` commands with specific flags, verify the flags exist locally (`ksail <command> --help`) before merging.
- When writing restore functionality with YAML round-trips, add a unit test that round-trips a real CRD YAML and verifies the output is non-empty and valid.
- Consider adding a pre-merge validation script that smoke-tests `ksail workload install --help` and checks for required flags.

## AI Team Self-Improvement

Add to `instructions.md` for coding agents:

```markdown
## workload install flags
When implementing or modifying `ksail workload install`, ensure these Helm flags are always present:
- `--version string` → `helm.ChartSpec.Version`
- `--timeout duration` → `helm.ChartSpec.Timeout`
- `--wait bool` → `helm.ChartSpec.Wait`
- `--atomic bool` → `helm.ChartSpec.Atomic`
- `--create-namespace bool` → `helm.ChartSpec.CreateNamespace`

## restore label injection
`injectRestoreLabels` in `pkg/cli/cmd/cluster/restore.go` uses YAML round-trip
(Unmarshal → SetLabels → Marshal). Any change to this function MUST be tested
with a real CRD YAML to ensure the output is non-empty and `apiVersion`/`kind`
are preserved. An empty output causes "no objects passed to apply" silently.
```

## Historical Context

No prior investigations found for these specific errors. Both issues were introduced/exposed by PR #3611 which added new E2E test coverage. The `workload install` flags gap is a pre-existing omission; the restore label injection bug may be pre-existing but was not tested until now.

> Generated by [CI Doctor](https://github.com/devantler-tech/ksail/actions/runs/23982711272/agentic_workflow) · ● 2.6M · [◷](https://github.com/search?q=repo%3Adevantler-tech%2Fksail+is%3Aissue+%22gh-aw-workflow-call-id%3A+devantler-tech%2Fksail%2Fci-doctor%22&type=issues)
>
> To install this [agentic workflow](https://github.com/githubnext/agentics/blob/1f672aef974f4246124860fc532f82fe8a93a57e/workflows/ci-doctor.md), run
> ```
> gh aw add githubnext/agentics/workflows/ci-doctor.md@1f672aef974f4246124860fc532f82fe8a93a57e
> ```

### PR DESCRIPTION
**Pull request by devantler**
Fixes two E2E test failures from PR #3611 (issue #3631).

**Missing install flags**: Added --version (default: latest) and --timeout (default: 5min) to workload install, wired to helm.ChartSpec.

**kubectl os.Exit bypass**: All embedded kubectl commands use Run with CheckErr which calls os.Exit(1). Added ExecuteSafely helper using BehaviorOnFatal+panic/recover. Applied to restore, backup, watch, edit, and resource_gen.

Fixes #3631

### CODE REVIEW

File: pkg/client/kubectl/execute.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+			if e, ok := r.(*kubectlFatalError); ok {
+				retErr = fmt.Errorf("%w", e)
+			} else {
+				// Re-panic for unexpected panics (bugs, nil pointers, etc.)
+				panic(r)
+			}
+		}
+	}()
+
+	return cmd.ExecuteContext(ctx)
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/execute.go]: Fixed. Added sync.Mutex (fatalMu) to serialize all BehaviorOnFatal overrides. Both ExecuteSafely and executeSafeRun share this mutex, making concurrent calls from errgroup safe.

File: pkg/cli/cmd/cluster/restore.go [hunk 2]
```diff
@@ -535,10 +535,12 @@ func restoreResourceFile(
...
-		return fmt.Errorf(
-			"kubectl failed: %w (output: %s)",
-			err, stderr,
-		)
+		errMsg := err.Error()
+		if stderr != "" {
+			errMsg = fmt.Sprintf("%s (output: %s)", errMsg, stderr)
+		}
+
+		return fmt.Errorf("kubectl failed: %s", errMsg)
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/restore.go]: Fixed. Restored %w wrapping to preserve error chain semantics.

File: .github/actions/ksail-system-test/action.yaml [hunk 3]
```diff
@@ -179,7 +262,82 @@ runs:
...
+        done
+        if [ "$apply_succeeded" != "true" ]; then
+          echo "❌ workload apply failed after 3 attempts"
+          exit 1
+        fi
+
+    - name: 🧪 ksail workload watch
+      id: workload-watch
+      if: inputs.apply-overlay-path != ''
+      continue-on-error: true
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/actions/ksail-system-test/action.yaml:283]: This change is from PR #3611 (E2E coverage improvements) that was merged into main. It is not part of this PR's scope. The continue-on-error was intentional for the watch test since it is inherently timing-sensitive.

File: pkg/client/kubectl/resource_gen.go [hunk 4]
```diff
@@ -330,10 +333,33 @@ func (c *Client) executeCommand(cmd *cobra.Command, args []string) error {
...
+// kubectl's cmdutil.CheckErr. This is necessary because kubectl commands
+// use Run (not RunE) with CheckErr which calls os.Exit on any error.
+func executeSafeRun(cmd *cobra.Command, args []string) (retErr error) {
+	cmdutil.BehaviorOnFatal(func(msg string, code int) {
+		panic(&kubectlFatalError{msg: msg, code: code})
+	})
+
+	defer func() {
+		cmdutil.DefaultBehaviorOnFatal()
+
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/resource_gen.go]: Fixed. executeSafeRun now shares the same fatalMu mutex with ExecuteSafely, serializing all BehaviorOnFatal overrides.

File: pkg/cli/cmd/cluster/backup.go [hunk 5]
```diff
@@ -504,7 +504,7 @@ func runKubectlGet(
 	getCmd.SilenceUsage = true
 	getCmd.SilenceErrors = true
 
-	err := getCmd.ExecuteContext(ctx)
+	err := kubectl.ExecuteSafely(ctx, getCmd)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/backup.go:508]: Fixed. ExecuteSafely is now guarded by fatalMu mutex, making concurrent calls from exportResources errgroup safe.

File: pkg/client/kubectl/resource_gen.go [hunk 6]
```diff
@@ -330,10 +333,37 @@ func (c *Client) executeCommand(cmd *cobra.Command, args []string) error {
 	}
 
 	if cmd.Run != nil {
-		cmd.Run(cmd, args)
-
-		return nil
+		return executeSafeRun(cmd, args)
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/resource_gen.go:341]: Fixed. Extracted a shared withSafeFatal helper in execute.go that both ExecuteSafely and executeSafeRun now delegate to. Single implementation, no duplication.

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -5,6 +5,13 @@ on:
     branches: [main]
   pull_request:
   merge_group:
+  workflow_dispatch:
+    inputs:
+      run_system_tests:
+        description: "Run system tests (normally only runs in merge queue)"
+        required: false
+        type: boolean
+        default: true
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/ci.yaml:14]: These CI changes (workflow_dispatch, gen-smoke-test job, etc.) are from PR #3611 (E2E coverage improvements) that was merged into this branch. They are not new work in this PR. They will be properly scoped when PR #3611 lands separately.

File: pkg/client/kubectl/execute.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+// process. This is essential when KSail calls kubectl commands internally
+// (backup, restore, watch) and needs to handle errors gracefully.
+func ExecuteSafely(ctx context.Context, cmd *cobra.Command) error {
+	if err := withSafeFatal(func() {
+		_ = cmd.ExecuteContext(ctx)
+	}); err != nil {
+		return err
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/execute.go]: Fixed. ExecuteSafely now captures and returns the ExecuteContext error in addition to any fatal error from BehaviorOnFatal.

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -432,11 +439,116 @@ jobs:
...
+          echo "=== Testing mcp --help ==="
+          ksail mcp --help
+          echo "✅ mcp --help passed"
+
   warm-helm-cache:
     name: 🔥 Warm Helm Cache
     runs-on: ubuntu-latest
     needs: [changes, build-artifact]
-    if: github.event_name == 'merge_group' && needs.changes.outputs.system-test == 'true'
+    if: (github.event_name == 'merge_group' && needs.changes.outputs.system-test == 'true') || (github.event_name == 'workflow_dispatch' && inputs.run_system_tests == true)
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/ci.yaml:586]: This CI workflow change is from PR #3611 (E2E coverage improvements), not this PR. It will be addressed in that PR's scope.

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -582,11 +694,102 @@
...
+        with:
+          go-version-file: go.mod
+          cache: false
+
+      - name: 📦 Cache KSail Binary
+        uses: ./.github/actions/cache-ksail-binary
+        with:
+          go-version: ${{ steps.setup-go.outputs.go-version }}
+          source-hash: ${{ hashFiles('go.mod', 'go.sum', '**/*.go') }}
+          output-path: /usr/local/bin/ksail
```
[the remark this answers is not in the document]
Comment (devantler) [reply .github/workflows/ci.yaml:793]: This zizmor alert is about CI workflow changes from PR #3611 (E2E coverage improvements) that were merged into this branch. It will be addressed in that PR's scope.

File: pkg/client/kubectl/resource_gen.go [hunk 11]
```diff
@@ -91,8 +91,11 @@ func (c *Client) CreatePriorityClassCmd() (*cobra.Command, error) {
 
 // newResourceCmd creates a gen command that wraps kubectl create with forced --dry-run=client -o yaml.
 func (c *Client) newResourceCmd(resourceType string) (*cobra.Command, error) {
-	// Use empty string for kubeconfig since --dry-run=client doesn't need cluster access
+	// Hold fatalMu during command construction: kubectl's NewCmdCreate calls
+	// CheckErr which reads the global fatalErrHandler without synchronization.
+	fatalMu.Lock()
 	tempCreateCmd := c.CreateCreateCommand("")
+	fatalMu.Unlock()
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/resource_gen.go]: Fixed. Extracted createCommandSafely helper that uses defer fatalMu.Unlock() for panic-safe release.

File: pkg/client/kubectl/resource_gen.go [hunk 12]
```diff
@@ -215,7 +218,12 @@ func (c *Client) executeSubcommandGen(
 	args []string,
 ) error {
 	freshClient := c.createFreshClient(cmd)
+
+	// Hold fatalMu during command construction: kubectl's NewCmdCreate calls
+	// CheckErr which reads the global fatalErrHandler.
+	fatalMu.Lock()
 	createCmd := freshClient.CreateCreateCommand("")
+	fatalMu.Unlock()
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/resource_gen.go]: Fixed. Extracted createCommandSafely helper that uses defer fatalMu.Unlock() for panic-safe release.

File: pkg/client/kubectl/resource_gen.go [hunk 13]
```diff
@@ -235,7 +243,12 @@ func (c *Client) executeSubcommandGen(
 // executeResourceGen executes kubectl create with forced --dry-run=client -o yaml flags.
 func (c *Client) executeResourceGen(resourceType string, cmd *cobra.Command, args []string) error {
 	freshClient := c.createFreshClient(cmd)
+
+	// Hold fatalMu during command construction: kubectl's NewCmdCreate calls
+	// CheckErr which reads the global fatalErrHandler.
+	fatalMu.Lock()
 	createCmd := freshClient.CreateCreateCommand("")
+	fatalMu.Unlock()
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/client/kubectl/resource_gen.go]: Fixed. Extracted createCommandSafely helper that uses defer fatalMu.Unlock() for panic-safe release.

File: pkg/cli/cmd/cluster/restore.go [hunk 14]
```diff
@@ -526,19 +526,29 @@ func restoreResourceFile(
...
 		stderr := errBuf.String()
 
+		// Empty YAML files (no resources of this type in the backup) are not
+		// an error — just skip silently.
+		if strings.Contains(err.Error(), "no objects passed to apply") ||
+			strings.Contains(err.Error(), "no objects passed to create") ||
+			strings.Contains(stderr, "no objects passed to apply") ||
+			strings.Contains(stderr, "no objects passed to create") {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/cli/cmd/cluster/restore.go]: Fixed. Added `isEmptyYAML` pre-check that detects files with only whitespace/separators and skips before invoking kubectl. Removed post-hoc string matching. Also extracted `classifyRestoreError` to reduce cyclomatic complexity.
